Creating friendly functions for establishing new controllers on existing or new fabrics

j-ororke opened the pull request:
- This is to help resolve task filed by Cecille here: [465](https://github.com/project-chip/matter-test-scripts/issues/465)
- Created 2 friendly functions for establishing new controllers on new and existing fabrics

#### Testing
Added standalone test script "src/python_testing/test_testing/TestCreateNewController.py" to validate

j-ororke committed:
Creating friendly functions to resolve issue 465:
- This is to help resolve task filed: https://github.com/project-chip/matter-test-scripts/issues/465
- Created 2 friendly functions for establishing new controllers on new and existing fabrics

restyled-commits committed:
Restyled by autopep8

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -1681,6 +1681,23 @@ def wait_for_user_input(self,
             logging.info("========= EOF on STDIN =========")
             return None
 
+    def create_new_controller_on_new_fabric(self, nodeid: int = 2, vendorid: bytes = 0xFFF1, fabricid: int = 1):
```
bytes for 0xFFF1 does not look like the correct type for vendorid. Something is mismatched here

j-ororke replied:
You are correct, thank you, I must have fat fingered when setting that value.
I have updated this in the latest code change.

andy31415 commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -1681,6 +1681,23 @@ def wait_for_user_input(self,
+        """
+        Create new controller on an already established fabric
+        """
+        current_fabric_admin = self.certificate_authority_manager.activeCaList[0].adminList[0]
```
aside for the comment above, I wonder about the syntax here: this is restricted to activeCaList[0] only (why are we not supporting other indices?) and also it seems to replace a one-liner (albeit a very long one liner....) like `self.certificate_authority_manager.activeCaList[0].adminList[0].NewController(nodeid)`

why do we place this on `MatterBaseTest` class? this is a class with over 1500 lines of code already ... that is not good for maintainability. How about having `certificate_authority_manager` have these methods? then instead of `self.new_controller(....)` I call `self.certificate_authority_manager.create_controller(....)` in some form of arguments. Likely we then control with activeCaList we use ... we could even have a union of existingCa vs NewCa.

cecille replied:
activeCaList[0] is where the default contoller is created. So 0 should be a default, but yeah, it would be good to be able to create ex. new controller on new fabric index 1, second controller on that newly created fabric.

j-ororke replied:
Please see following changes made to resolve the above comment:
- Set the activeCAList[] list value to now be dynamically set when calling the function with a default value set to 0
- Moved the code from the matter_testing support module to the CertificateAuthority module, can now be called as self.certificate_authority_manager.create_new_controller(....) 
- Created a new standalone test module to test the function added in this PR

Please let me know if these new changes will work, I am always happy to update and adjust if this is not what was wanted. Or even do some further changes here if wanted as well!

cecille commented on src/python_testing/matter_testing_infrastructure/chip/testing/matter_testing.py:
```diff
@@ -1681,6 +1681,23 @@ def wait_for_user_input(self,
+        new_controller = new_fabric_admin.NewController(nodeId=nodeid)
+        return new_controller
+
+    def create_new_controller_on_existing_fabric(self, nodeid: int = 2):
```
also fine. I didn't think too hard about it when I created the issue other than "this code is yucky, let's make it less yucky in some way"

j-ororke replied:
I have updated the latest code to using just one function for both operations of if using a new or existing fabric as mentioned above. I think that works better and it definitely feels cleaner and should be a bit easier to maintain if the code is contained inside of one single function.

j-ororke committed:
Updating to resolve comments:
- Moved code from matter_testing support module to CertificateAuthority module's CertificateAuthorityManager class
- Changed to using 1 single function for both creating controllers, thne if statement to determine if using a new or existing fabric
- Created new standalone script to test functionality.

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating TC_CreateNewController python test module:
- Resolving linting errors

j-ororke committed:
Update CertificateAuthority.py

Changiing adminList[] value back to 0 for the moment.

andy31415 commented on src/controller/python/chip/CertificateAuthority.py:
```diff
@@ -315,6 +315,21 @@ def NewCertificateAuthority(self, caIndex: Optional[int] = None, maximizeCertCha
 
         return ca
 
+    def create_new_controller(self, vendorid: int = 0xFFF1, fabricId: int = 1, new_fabric: bool = True, CaList: int = 0, nodeid: int = 2):
```
`CaList` as a name does not match the others. Overall we have to pick a style as we seem to have:
   - vendrid and nodeid are `lowercaseconcatenated`
   - fabricId is `lowerCamelCase`
   - new_fabric is `lower_snake_case`
   - CaList is `UppserCamelCase`
   
 I like consistency. Let's use `lower_camel_case` for all argument names as I believe this is typical for python.

j-ororke replied:
That sounds good, 
It is a bit interesting that the CaList arg name triggered the git syntax coloring in that way lol. 
I will get this updated to using lower_camel_case for the argument names.

j-ororke committed:
Updating CertificateAuthority and TC_CreateNewController modules:
- Updated to change arg names to using lower_snake_case for create_new_controller function.

cecille commented on src/python_testing/TC_CreateNewController.py:
```diff
@@ -0,0 +1,128 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
```
```suggestion
#    Copyright (c) 2025 Project CHIP Authors
```

cecille commented on src/python_testing/TC_CreateNewController.py:
```diff
@@ -0,0 +1,128 @@
+#
+#    Copyright (c) 2023 Project CHIP Authors
```
Would you mind changing the name of this file to TestCreateNewController.py so it doesn't look like a cert test?

j-ororke replied:
No problem, I will get the name of the script changed

cecille commented on src/controller/python/chip/CertificateAuthority.py:
```diff
@@ -315,6 +315,21 @@ def NewCertificateAuthority(self, caIndex: Optional[int] = None, maximizeCertCha
 
         return ca
 
+    def create_new_controller(self, vendor_Id: int = 0xFFF1, fabric_Id: int = 1, new_Fabric: bool = True, ca_List: int = 0, node_Id: int = 2):
```
I'm not totally sure this is right place for this - this is now part of the certificate authority manager class, but it's creating a controller. I think this might be better off as a standalone helper function somewhere.

@andy31415 - thoughts? I'm open to being overruled.

andy31415 replied:
Also `caList` is just a member accessor ... so no need for a method really.

andy31415 replied:
Apparently my old reqest was to have code like:

```py
@dataclass
class NewFabricInfo:
    vendorId: int = 0xFFF1
    fabricId: int = 1

def new_controller(self, nodeid: int = 2, new_fabric: Optional[NewFabricInfo] = None):
    if new_fabric is not None:
        fabric_admin = ....
    else:
        fabric_admin = ...
    return fabric_admin.NewController(nodeId=nodeId)
```

This was not done really like this, but rather with many extra arguments (so now we have a bool and unused arguments).

andy31415 replied:
Maybe what I am suggesting then would be:

```py
def new_fabric_admin(self, vendorId: int, fabricId: int):
   # ....
   return ....
```

Would that be usable? then I do not do the new-controller inside ca admin

j-ororke replied:
Updated

j-ororke committed:
Update src/python_testing/TC_CreateNewController.py

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
Updating tests.yaml and created TestCreateNewController.py:
- Added TestCreateNewController.py to run during CI
- Renamed TC_CreateNewController.py to TestCreateNewController.py
- Changed setup_class() in TestCreateNewController to get it working after name change

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updated test_metadata script:
- Added TestCreateNewControllerScript.py to make sure it does not run in tests during CI checks

j-ororke committed:
Merge remote-tracking branch 'upstream/master' into add_friendly_funct_fabcon

j-ororke committed:
Merge branch 'master' into add_friendly_funct_fabcon

restyled-commits committed:
Restyled by ruff

j-ororke committed:
Updating code to match suggestion from Andrei

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

andy31415 commented:
TODO: @andy31415  to review.
